spec/integration/rails/commit_hooks_spec.rb

```diff
@@ -0,0 +1,406 @@
+            raise 'rollitback'
+          end
+        end
+
+        after_commit only: :create do |employee|
+          Callbacks.add(:employee_after_create, employee)
+        end
+
+        after_commit do |employee|
+          Callbacks.add(:employee_after_create_eval_test, self)
+        end
+
+        has_many :positions
+      end
+    end
+
+    controller(ApplicationController) do
+      def create
+        employee = resource.build(params)
+
+        if employee.save
+          render jsonapi: employee
+        else
+          raise 'whoops'
+        end
+      end
+
+      def update
+        employee = resource.find(params)
+
+        if employee.update_attributes
+          render_jsonapi(employee, scope: false)
+        else
+          raise 'whoops'
+        end
+      end
+
+      def destroy
+        employee = resource.find(params)
+
+        if employee.destroy
+          render jsonapi: employee
+        else
+          raise 'whoops'
+        end
+      end
+
+      def resource
+        IntegrationHooks::EmployeeResource
+      end
+
+      private
+
+      def params
+        @params ||= begin
+          hash = super.to_unsafe_h.with_indifferent_access
+          hash = hash[:params] if hash.has_key?(:params)
+          hash
+        end
+      end
+    end
+
+    before do
+      @request.headers['Accept'] = Mime[:json]
+      @request.headers['Content-Type'] = Mime[:json].to_s
+
+      routes.draw {
+        post "create" => "anonymous#create"
+        put "update" => "anonymous#update"
+        delete "destroy" => "anonymous#destroy"
+      }
+    end
+
+    def json
+      JSON.parse(response.body)
+    end
+
+    let(:payload) do
+      {
+        data: {
+          type: 'employees',
+          attributes: { first_name: 'Jane' }
+        }
+      }
+    end
+
+    context 'on create' do
+      it 'fires after validations but before ending the transaction' do
+        expect_any_instance_of(Graphiti::Util::ValidationResponse)
+          .to receive(:validate!)
+        expect {
+          post :create, params: payload
+        }.to raise_error('rollitback')
+        expect(Employee.count).to be_zero
+        expect(Callbacks.entities.length).to eq(1)
+        expect(Callbacks.fired[:create]).to be_a(Employee)
+      end
+
+      context 'when an error is raised before_commit' do
+        it 'does not run after_commit callbacks' do
+          expect {
+            post :create, params: payload
+          }.to raise_error('rollitback')
+          expect(Callbacks.entities).not_to include(:employee_after_create)
+        end
+      end
+
+      context 'when validation fails' do
+        let(:validation_response) do
+          double('Graphiti::Util::ValidationResponse').as_null_object
+        end
+
+        before do
+          allow_any_instance_of(Graphiti::Util::ValidationResponse)
+            .to receive(:validate!).and_raise(Graphiti::Errors::ValidationError, validation_response)
+        end
+        j
```

Comment:
The letter `j` is randomly here? 😆

Reply:
yeah i was just checking your code review skills...